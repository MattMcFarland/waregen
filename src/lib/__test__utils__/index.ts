import { clone, range } from "@@/utils";
import { DefaultWareEntity, DefaultBlueprintEntity } from "@@/entities/Ware";
import { AddwareEntity } from "@@/XMLTypes/X4WareGenXML";
import { GeneratorConfig } from "@@/generator";

export const mockGeneratorConfig = (): GeneratorConfig => ({
  defaultWare: defaultWare(),
  defaultBlueprint: defaultBlueprint(),
  addwaresList: mockAddwares(4),
  gamePath: "/test/Game",
  modPath: "extensions/test",
  modPrefix: "pfx",
  unpackedPath: "unpacked"
});

export const mockAddwares = (count: number): AddwareEntity[] => {
  const char = (code: number) => "abcdefghijklmnopqrstuvwxyz"[code];

  return range(count).map(
    (i: number): AddwareEntity => ({
      Attributes: {
        baskets: `[test_basket_${char(i)}]`,
        cloneProductionModuleFrom: `prod_cloned_macro_${char(i)}`,
        id: `${char(i)}`
      },
      ware: [mockWare(`${char(i)}`)],
      blueprint: [mockBlueprint(`${char(i)}`)]
    })
  );
};

export const mockWare = (name: string) => {
  const ware = defaultWare();
  ware.Attributes.name = name;
  return clone(ware);
};
export const mockBlueprint = (name: string) => {
  const bp = defaultBlueprint();
  bp.Attributes.name = `${name} production`;
  return clone(bp);
};

export const defaultBlueprint = (): DefaultBlueprintEntity => {
  const bp = <DefaultBlueprintEntity>defaultWare();
  bp.Attributes.tags = "module";
  bp.Attributes.description = "blueprint description";
  bp.restriction = [
    {
      Attributes: { licence: "station_gen_basic " }
    }
  ];
  bp.research = [
    {
      Attributes: { time: 10 },
      research: [
        { ware: [{ Attributes: { ware: "research_module_production" } }] }
      ]
    }
  ];
  bp.owner = [
    { Attributes: { faction: "antigone" } },
    { Attributes: { faction: "argon" } },
    { Attributes: { faction: "holyorder" } },
    { Attributes: { faction: "paranid" } },
    { Attributes: { faction: "teladi" } }
  ];
  bp.production = [
    {
      Attributes: {
        time: 791,
        amount: 1,
        method: "default",
        name: "{20206,101}"
      },
      primary: [
        {
          ware: [
            { Attributes: { ware: "claytronics", amount: 30 } },
            { Attributes: { ware: "energycells", amount: 75 } },
            { Attributes: { ware: "hullparts", amount: 150 } }
          ]
        }
      ]
    }
  ];
  return clone(bp);
};

export const defaultWare = (): DefaultWareEntity =>
  clone({
    Attributes: {
      id: "default",
      name: "default",
      transport: "container",
      volume: 1
    },
    price: [
      {
        Attributes: {
          min: 1,
          average: 1,
          max: 1
        }
      }
    ],
    production: [
      {
        Attributes: {
          time: 10,
          amount: 1,
          method: "default",
          name: "{20206, 101}"
        },
        primary: [
          {
            ware: [{ Attributes: { ware: "energycells", amount: 50 } }]
          }
        ],
        effects: [
          {
            effect: [
              {
                Attributes: { type: "efficiency", product: 1 }
              }
            ]
          }
        ]
      }
    ]
  });
