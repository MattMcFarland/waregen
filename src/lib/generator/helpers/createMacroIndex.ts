import { IndexHash } from "../../entities";
import { IndexByString } from "../../utils/object/types";
import { PathBuilder } from "../../utils/PathBuilder";
import { GeneratorConfig } from "../../generator";
import { IdRoster } from "../../utils/IdRoster";

export default function createMacroIndex(config: GeneratorConfig): IndexHash {
  const hash: IndexByString = config.addwaresList.reduce(
    (acc: IndexByString, addware) => {
      const baseId = addware.Attributes.id;
      const ids = new IdRoster(baseId, config.modPrefix);

      acc[ids.productionMacro] = new PathBuilder(config, baseId)
        .dirModPath()
        .fileProductionMacroXML()
        .toString();
      acc[ids.wareMacro] = new PathBuilder(config, baseId)
        .dirModPath()
        .fileWareMacroXML()
        .toString();
      return acc;
    },
    {}
  );

  return new IndexHash(hash);
}
