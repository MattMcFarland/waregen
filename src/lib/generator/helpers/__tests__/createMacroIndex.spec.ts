import { mockGeneratorConfig } from "@@/__test__utils__";
import createMacroIndex from "../createMacroIndex";

describe("generator/helper", () => {
  describe("createMacroIndex", () => {
    test("toJson()", () => {
      const mockConfig = mockGeneratorConfig();
      const macroIndex = createMacroIndex(mockConfig);
      expect(macroIndex.toJson()).toMatchSnapshot();
    });
    test("toXml()", () => {
      const mockConfig = mockGeneratorConfig();
      const macroIndex = createMacroIndex(mockConfig);
      expect(macroIndex.toXml()).toMatchSnapshot();
    });
  });
});
