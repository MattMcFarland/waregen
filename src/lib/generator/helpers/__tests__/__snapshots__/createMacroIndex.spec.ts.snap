// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator/helper createMacroIndex toJson() 1`] = `"{\\"entry\\":[{\\"Attributes\\":{\\"name\\":\\"pfx_prod_gen_ware0_macro\\",\\"value\\":\\"extensions\\\\\\\\test\\\\\\\\assets\\\\\\\\structures\\\\\\\\production\\\\\\\\macros\\\\\\\\pfx_prod_gen_ware0_macro.xml\\"}},{\\"Attributes\\":{\\"name\\":\\"pfx_prod_gen_ware1_macro\\",\\"value\\":\\"extensions\\\\\\\\test\\\\\\\\assets\\\\\\\\structures\\\\\\\\production\\\\\\\\macros\\\\\\\\pfx_prod_gen_ware1_macro.xml\\"}},{\\"Attributes\\":{\\"name\\":\\"pfx_prod_gen_ware2_macro\\",\\"value\\":\\"extensions\\\\\\\\test\\\\\\\\assets\\\\\\\\structures\\\\\\\\production\\\\\\\\macros\\\\\\\\pfx_prod_gen_ware2_macro.xml\\"}},{\\"Attributes\\":{\\"name\\":\\"pfx_prod_gen_ware3_macro\\",\\"value\\":\\"extensions\\\\\\\\test\\\\\\\\assets\\\\\\\\structures\\\\\\\\production\\\\\\\\macros\\\\\\\\pfx_prod_gen_ware3_macro.xml\\"}},{\\"Attributes\\":{\\"name\\":\\"pfx_prod_gen_ware4_macro\\",\\"value\\":\\"extensions\\\\\\\\test\\\\\\\\assets\\\\\\\\structures\\\\\\\\production\\\\\\\\macros\\\\\\\\pfx_prod_gen_ware4_macro.xml\\"}}]}"`;

exports[`generator/helper createMacroIndex toXml() 1`] = `
"<index>
  <entry name=\\"pfx_prod_gen_ware0_macro\\" value=\\"extensions\\\\test\\\\assets\\\\structures\\\\production\\\\macros\\\\pfx_prod_gen_ware0_macro.xml\\"/>
  <entry name=\\"pfx_prod_gen_ware1_macro\\" value=\\"extensions\\\\test\\\\assets\\\\structures\\\\production\\\\macros\\\\pfx_prod_gen_ware1_macro.xml\\"/>
  <entry name=\\"pfx_prod_gen_ware2_macro\\" value=\\"extensions\\\\test\\\\assets\\\\structures\\\\production\\\\macros\\\\pfx_prod_gen_ware2_macro.xml\\"/>
  <entry name=\\"pfx_prod_gen_ware3_macro\\" value=\\"extensions\\\\test\\\\assets\\\\structures\\\\production\\\\macros\\\\pfx_prod_gen_ware3_macro.xml\\"/>
  <entry name=\\"pfx_prod_gen_ware4_macro\\" value=\\"extensions\\\\test\\\\assets\\\\structures\\\\production\\\\macros\\\\pfx_prod_gen_ware4_macro.xml\\"/>
</index>"
`;
